cylc/flow/scripts/profiler.py

```diff
@@ -0,0 +1,187 @@
+        with open(cgroup_cpu_path, 'r') as f:
+            for line in f:
+                # Cgroups v2 uses nanoseconds
+                return int(line) / 1000000
+
+
+def write_data(data, filename):
+    with open(filename, 'w') as f:
+        f.write(data + "\n")
+
+
+def get_cgroup_version(cgroup_location: str, cgroup_name: str) -> int:
+    # HPC uses cgroups v2 and SPICE uses cgroups v1
+    if Path.exists(Path(cgroup_location + cgroup_name)):
+        return 1
+    elif Path.exists(Path(cgroup_location + "/memory" + cgroup_name)):
+        return 2
+    else:
+        raise FileNotFoundError("Cgroup not found")
+
+
+def get_cgroup_name():
+    """Get the cgroup directory for the current process"""
+    # Get the PID of the current process
+    pid = os.getpid()
+    try:
+        # Get the cgroup information for the current process
+        with open('/proc/' + str(pid) + '/cgroup', 'r') as f:
+            result = f.read()
+        result = PID_REGEX.search(result).group()
+        return result
+    except FileNotFoundError as err:
+        raise FileNotFoundError(
+            '/proc/' + str(pid) + '/cgroup not found') from err
+
+    except AttributeError as err:
+        raise AttributeError("No cgroup found for process:", pid) from err
+
+
+def get_cgroup_paths(version, location, name):
+
+    if version == 1:
+        return Process(
+            cgroup_memory_path=location +
+            name + "/" + "memory.peak",
+            cgroup_cpu_path=location +
+            name + "/" + "cpu.stat")
+
+    elif version == 2:
+        return Process(
+            cgroup_memory_path=location + "/memory" +
+            name + "/memory.max_usage_in_bytes",
+            cgroup_cpu_path=location + "/cpu" +
+            name + "/cpuacct.usage")
+
+
+def profile(process, version, delay, keep_looping=lambda: True):
+    # The infinite loop that will constantly poll the cgroup
+    # The lambda function is used to allow the loop to be stopped in unit tests
+    peak_memory = 0
+    while keep_looping():
+        # Write cpu / memory usage data to disk
+        cpu_time = parse_cpu_file(process.cgroup_cpu_path, version)
+        write_data(str(cpu_time), "cpu_time")
+
+        memory = parse_memory_file(process.cgroup_memory_path)
+        # Only save Max RSS to disk if it is above the previous value
+        if memory > peak_memory:
+            peak_memory = memory
+            write_data(str(peak_memory), "max_rss")
+
+        time.sleep(delay)
+
+
+def get_config(args):
+    # Find the cgroup that this process is running in.
+    # Cylc will put this profiler in the same cgroup
+    # as the job it is profiling
+    cgroup_name = get_cgroup_name()
+    cgroup_version = get_cgroup_version(args.cgroup_location, cgroup_name)
+    process = get_cgroup_paths(cgroup_version,
+                               args.cgroups_location,
+                               cgroup_name)
+
+    profile(process, cgroup_version, args.delay)
```

Comment:
Codecov has spotted a couple of test coverage holes here that it would be good to fill.